pkg/kube/dialer.go

```diff
@@ -0,0 +1,336 @@
+package kube
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/transport/spdy"
+)
+
+const (
+	defaultClusterDomain = "cluster.local"
+	defaultNamespace     = "default"
+	svcLabel             = "svc"
+
+	portForwardProtocolV1Name = "portforward.k8s.io"
+)
+
+var (
+	ErrInvalidPodFQDN = errors.New("invalid pod FQDN")
+	ErrNoPort         = errors.New("no port specified")
+
+	negotiatedSerializer = serializer.NewCodecFactory(runtime.NewScheme()).WithoutConversion()
+)
+
+// PodDialer is a basic port-forwarding dialer that doesn't start
+// any local listeners, but returns a net.Conn directly.
+type PodDialer struct {
+	config        *rest.Config
+	clusterDomain string
+	requestID     int
+
+	connections map[types.NamespacedName]httpstream.Connection
+	mutex       sync.RWMutex
+}
+
+// NewPodDialer create a PodDialer.
+func NewPodDialer(config *rest.Config) *PodDialer {
+	return &PodDialer{
+		config:        config,
+		clusterDomain: defaultClusterDomain,
+		connections:   make(map[types.NamespacedName]httpstream.Connection),
+	}
+}
+
+// WithClusterDomain overrides the domain used for FQDN parsing.
+func (p *PodDialer) WithClusterDomain(domain string) *PodDialer {
+	p.clusterDomain = domain
+	return p
+}
+
+// Dial dials the given pod's service-based DNS address and returns a
+// net.Conn that can be used to reach the pod directly.
+func (p *PodDialer) Dial(network string, address string) (net.Conn, error) {
+	switch network {
+	case "tcp", "tcp4", "tcp6":
+	default:
+		return nil, fmt.Errorf("dialer only supports TCP-based networks: %w", net.UnknownNetworkError(network))
+	}
+
+	pod, port, err := p.parseDNS(address)
+	if err != nil {
+		return nil, err
+	}
+
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	conn, err := p.connectionForPod(pod)
+	if err != nil {
+		return nil, err
+	}
+
+	p.requestID++
+
+	headers := http.Header{}
+	headers.Set(corev1.PortHeader, strconv.Itoa(port))
+	headers.Set(corev1.PortForwardRequestIDHeader, strconv.Itoa(p.requestID))
+
+	headers.Set(corev1.StreamType, corev1.StreamTypeError)
+	errorStream, err := conn.CreateStream(headers)
+	if err != nil {
+		return nil, err
+	}
+
+	headers.Set(corev1.StreamType, corev1.StreamTypeData)
+	dataStream, err := conn.CreateStream(headers)
+	if err != nil {
+		return nil, err
+	}
+
+	return wrapConn(network, address, dataStream, errorStream), nil
+}
+
+// DialContext is a simple wrapper around Dial.
+func (p *PodDialer) DialContext(_ context.Context, network string, address string) (net.Conn, error) {
+	return p.Dial(network, address)
+}
+
+// parseDNS attempts to determine the intended pod to target, currently the
```

Comment:
This is a beautiful comment! The symmetry of dialing within a k8s cluster is nice but I think it's too tailored towards StatefulSets, it introduces a needless information overhead, and the defaulting semantics could very easily get me in trouble.

Let's say I want to dial into a core DNS Pod so I can resolve the `kubernetes` service to find the cluster domain. I won't have a cluster domain as I'm trying to fetch it, the supported formats aren't valid hostnames within the cluster as core DNS isn't run under a headless service, and the namespace I'm dialing into (`kube-system`) is different from the default (Assuming that I'll be dialing into a different namespace later once I have the hostname).

I'd recommend only supporting `[pod-name].[namespace-name]`.

If we need to support the FQDN to make hostname verification work correctly, require the format to be `[pod-name].[service-name].[namespace-name].svc[.cluster-domain][.]` (cluster domain and trailing `.` would be optional). Ignore the validation of `service-name` and `cluster-domain` as hostname verification will handle that for you. This will remove any possible ambiguity between the two formats and ensure no defaulting takes place accidentally.